python/repositories.bzl

```diff
@@ -1,8 +1,164 @@
+            quiet = True,
+            working_directory = working_directory,
+        )
+        zstd = "{working_directory}/zstd".format(working_directory = working_directory)
+        unzstd = "./unzstd"
+        rctx.symlink(zstd, unzstd)
+
+    exec_result = rctx.execute([
+        "tar",
+        "--extract",
+        "--strip-components=2",
+        "--use-compress-program={unzstd}".format(unzstd = unzstd),
+        "--file={}".format(release_filename),
+    ])
+    if exec_result.return_code:
+        fail(exec_result.stderr)
+
+    build_content = """\
+# Generated by python/repositories.bzl
+
+load("@bazel_tools//tools/python:toolchain.bzl", "py_runtime_pair")
+
+filegroup(
+    name = "files",
+    srcs = glob(["bin/**", "lib/**", "include/**", "share/**"]),
+)
+
+exports_files(["bin/python3"])
+
+py_runtime(
+    name = "py2_runtime",
+    interpreter_path = "/bin/false",
+    python_version = "PY2",
+)
+
+py_runtime(
+    name = "py3_runtime",
+    interpreter = "bin/python3",
+    python_version = "PY3",
+)
+
+py_runtime_pair(
+    name = "python_runtimes",
+    py2_runtime = ":py2_runtime",
+    py3_runtime = ":py3_runtime",
+)
+"""
+    rctx.file("BUILD.bazel", build_content)
+
+python_repositories = repository_rule(
+    _python_repo_impl,
+    doc = _DOC,
+    attrs = dict({
+        "_zstd_sha256": attr.string(
+            default = "7c42d56fac126929a6a85dbc73ff1db2411d04f104fae9bdea51305663a83fd0",
+        ),
+        "_zstd_url": attr.string(
+            default = "https://github.com/facebook/zstd/releases/download/v{version}/zstd-{version}.tar.gz",
+        ),
+        "_zstd_version": attr.string(
+            default = "1.5.2",
+        ),
+    }, **_ATTRS),
+)
+
+# Wrapper macro around everything above, this is the primary API.
+def python_register_toolchains(name, **kwargs):
+    """Convenience macro for users which does typical setup.
+
+    - Create a repository for each built-in platform like "python_linux_amd64" -
+      this repository is lazily fetched when Python is needed for that platform.
+    - TODO(f0rmiga): create a convenience repository for the host platform like
+      "python_host".
+    - Create a repository exposing toolchains for each platform like
+      "python_platforms".
+    - Register a toolchain pointing at each platform.
+    Users can avoid this macro and do these steps themselves, if they want more
+    control.
+    Args:
+        name: base name for all created repos, like "python38".
+        **kwargs: passed to each python_repositories call.
+    """
+    for platform in PLATFORMS.keys():
+        python_repositories(
+            name = "{name}_{platform}".format(
+                name = name,
+                platform = platform,
+            ),
+            platform = platform,
+            **kwargs
+        )
+        native.register_toolchains("@{name}_toolchains//:{platform}_toolchain".format(
+            name = name,
+            platform = platform,
+        ))
```

Comment:
How does this work with `pip_parse`. As far as I remember, it is impossible to pass something with a `select`. Would having a single python_repository be preferable?

Reply:
No, the native toolchain system cannot be used in repository rules. We have to pass the interpreter explicitly to pip_parse/pip_install.

Reply:
Yes, I understand that, I probably did not find a better place to comment on that, but it would be great to have an explicit example on how we would be using such platform specific python interpreter repository in a `pip_parse` rule, but maybe I am missing something.

From my experience a different system python version may lead to incompatible Python byte-code at runtime. For example, a user is using an old Mac OS X version and the system python is `3.8`, but the repo is using `3.9`, then the test targets would fail if there is a minor version mismatch. To make things worse (or better), the default Python used by `pip_parse` cannot be upgraded by doing `brew install` (at least by default).

As long as we cannot have a consistent way to use the same Python to do `pip_parse` and as a toolchain, the usefulness may be slightly limited. The only way I found it possible to use together with `pip_parse` is by having a single repository rule that does `uname -r` and downloads an appropriate binary and exposing it as `@interpreter//:bin/python`, which can then be passed to `pip_parse` via the `python_interpreter_target` attribute.

Or maybe we should handle the this (i.e. selecting the correct python interpreter) within `pip_parse`?

Reply:
See https://github.com/bazelbuild/rules_python/pull/618/commits/d12143673d10660387a9bfdd87bfca1a55c9e0bf.